### PR TITLE
Drain unnecessary middle levels

### PR DESCRIPTION
**Pull request by ajkr**
An internal user asked about migrating from non-dynamic to dynamic level compaction, and the answer was unfortunately the nonempty levels that become unnecessary after the migration will never be automatically drained. Draining levels is also useful for dynamically changing `max_bytes_for_level_multiplier` to be effective, or to reduce read-amp after users delete data and the DB shrinks.

This PR:

- Finds levels to drain that are nonempty and above the ideal base level. Those levels are assigned target size 0.
- Levels with target size 0 have their compaction score computed differently to ensure it's always at least 1.0.

Test Plan:

- `make check -j64`
- use db_bench to verify levels are emptied after migration from non-dynamic to dynamic. Before this PR, running the below commands caused all of L0-L6 to contain data. After this PR only L0, L4-L6 contain data.
```
$ TEST_TMPDIR=/dev/shm ./db_bench -benchmarks=fillrandom  -num=1000000 -write_buffer_size=1048576 -target_file_size_base=1048576 -max_bytes_for_level_base=4194304 -compression_type=none -max_background_jobs=12 -benchmark_write_rate_limit=1048576
$ TEST_TMPDIR=/dev/shm ./db_bench -benchmarks=overwrite -use_existing_db=true -num=1000000 -write_buffer_size=1048576 -target_file_size_base=1048576 -max_bytes_for_level_base=4194304 -compression_type=none -max_background_jobs=12 -benchmark_write_rate_limit=1048576 -level_compaction_dynamic_level_bytes=1
```

### CODE REVIEW

**Comment by gwicke** [comment]
Thank you @ajkr. This feature would be very useful to us in at least two ways:

- We have use cases where data size per DB decreases significantly over time. Clearing out extra levels will avoid performance degradations, excessive flash burn and space usage.
- Migrations to dynamic level sizing currently require a full compaction, with writes disabled. With this feature, migration could become fairly automatic.

File: db/version_set.cc [hunk 1]
```diff
@@ -1640,8 +1640,16 @@ void VersionStorageInfo::ComputeCompactionScore(
...
       }
-      score = static_cast<double>(level_bytes_no_compacting) /
-              MaxBytesForLevel(level);
+      uint64_t max_bytes = MaxBytesForLevel(level);
+      if (max_bytes == 0) {
+        score = static_cast<double>(level_bytes_no_compacting) /
+                mutable_cf_options.max_bytes_for_level_base;
+        // This level is no longer necessary so make sure it is always eligible
+        // for compaction.
+        score = std::max(score, 1.0);
```
Comment (gwicke) [inline db/version_set.cc]: If we want to prioritize cleaning up higher levels, we could vary the minimum score. Example:
```cpp
score = std::max(score, 1.0 + static_cast<double>(MaxInputLevel() - level) / 10);
```

File: db/version_set.cc [hunk 2]
```diff
@@ -2510,17 +2511,31 @@ void VersionStorageInfo::CalculateBaseBytes(const ImmutableCFOptions& ioptions,
         }
       }
 
+      const uint64_t kMinNonEmptyLevelSize = static_cast<uint64_t>(
+          base_bytes_max / options.max_bytes_for_level_multiplier);
```
Comment (siying) [inline db/version_set.cc]: This is only the naming for "Variables declared constexpr or const, and whose value is fixed for the duration of the program": https://google.github.io/styleguide/cppguide.html#Constant_Names . This is probably doesn't qualify.

File: db/version_set.cc (hunk 1 above)
Comment (siying) [reply db/version_set.cc]: I agree that 1.0 maybe a little bit small. We don't want to fir compact L4->L5 while we are still accumulating data L0->L3, and we are going to compact L3->L4 anyway. I would say something like 1.5 or so feels OK to me, if we don't want to make things too complicated.

**Comment by ajkr** [comment]
I wonder if we land this can we get rid of the "More existing levels in DB than needed. max_bytes_for_level_multiplier may not be guaranteed." warning? With this PR RocksDB can automatically fix an excess number of levels, so there's nothing to alert the user about.

**Comment by miasantreble** [comment]
@ajkr can you please rebase to resolve conflicts and then we can land this. Thanks

**Comment by siying** [comment]
@ajkr I totally forgot the context of this PR. What's your suggestion?

**Comment by ajkr** [comment]
#11340 replaces this

**Pull request closed without merging**